File: static/java/src/main/java/com/epi/HashDictionary.java
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

package com.epi;

import java.util.Random;

public class HashDictionary {
  private static String randString(int len) {
    StringBuilder ret = new StringBuilder();
    Random rnd = new Random();

    while (len-- > 0) {
      ret.append((char)(rnd.nextInt(26) + 97));
    }
    return ret.toString();
  }

  // @include
  public static int stringHash(String str, int modulus) {
    int kMult = 997;
    int val = 0;
    for (int i = 0; i < str.length(); i++) {
      char c = str.charAt(i);
      val = (val * kMult + c) % modulus;
    }
    return val;
  }
  // @exclude

  public static void main(String[] args) {
    String str = null;
    if (args.length == 1) {
      str = args[0];
    } else {
      Random rnd = new Random();
      str = randString(rnd.nextInt(20) + 1);
    }
    System.out.println("string = " + str);
    System.out.println(stringHash(str, 1 << 16));
  }
}

File: static/solutions/cpp/Hash_dictionary.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <iostream>
#include <numeric>
#include <random>
#include <string>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::string;
using std::uniform_int_distribution;

string RandString(int len) {
  string ret;
  default_random_engine gen((random_device())());
  while (len--) {
    uniform_int_distribution<int> zero_or_one(0, 1);
    uniform_int_distribution<char> dis(0, 26);
    ret += (zero_or_one(gen) ? 'a' : 'A') + dis(gen);
  }
  return ret;
}

// @include
int StringHash(const string& str, int modulus) {
  const int kMult = 997;
  // @exclude
  /*
  return accumulate(str.cbegin(), str.cend(), 0,
    [kMult, modulus](const int &val, const char &c) {
      return (val * kMult + c) % modulus;
    });
  */
  // @include
  int val = 0;
  for (char c : str) {
    val = (val * kMult + c) % modulus;
  }
  return val;
}
// @exclude

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  string str;
  if (argc == 2) {
    str = argv[1];
  } else {
    uniform_int_distribution<int> dis(1, 20);
    str = RandString(dis(gen));
  }
  cout << "string = " << str << endl;
  cout << StringHash(str, 1 << (16)) << endl;
  return 0;
}
